### PR TITLE
[Notifier] Add Primotexto bridge

### PR DESCRIPTION
**Pull request by Goazil**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | none
| License       | MIT

This PR adds a new bridge for [Primotexto](https://primotexto.com/). Based on documentation available here: https://www.primotexto.com/

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -2814,6 +2814,7 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
             NotifierBridge\Yunpian\YunpianTransportFactory::class => 'notifier.transport_factory.yunpian',
             NotifierBridge\Zendesk\ZendeskTransportFactory::class => 'notifier.transport_factory.zendesk',
             NotifierBridge\Zulip\ZulipTransportFactory::class => 'notifier.transport_factory.zulip',
+            NotifierBridge\Primotexto\PrimotextoTransportFactory::class => 'notifier.transport_factory.primotexto',
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Please alpha sort

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.php [hunk 2]
```diff
@@ -112,6 +112,7 @@
         'unifonic' => Bridge\Unifonic\UnifonicTransportFactory::class,
         'vonage' => Bridge\Vonage\VonageTransportFactory::class,
         'yunpian' => Bridge\Yunpian\YunpianTransportFactory::class,
+        'primotexto' => Bridge\Primotexto\PrimotextoTransportFactory::class,
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.php]: Please alpha sort

File: src/Symfony/Component/Notifier/Bridge/Primotexto/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.1
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/CHANGELOG.md]:
```suggestion
7.2
```

File: src/Symfony/Component/Notifier/Bridge/Primotexto/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+---
+
+* Add the bridge
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/CHANGELOG.md]:
```suggestion
 * Add the bridge
```

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+use Symfony\Component\Notifier\Message\MessageOptionsInterface;
+
+/**
+ * @author Samaël tomas <samael.tomas@gmail.com>
+ */
+final class PrimotextoOptions implements MessageOptionsInterface
+{
+    private array $options;
+
+    public function __construct(array $options = [])
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php]:
```suggestion
    public function __construct(private array $options = [])
```
And linebreak the

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoTransport.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ */
+final class PrimotextoTransport extends AbstractTransport
+{
+    protected const HOST = 'api.primotexto.com';
+
+    private string $apiKey;
+    private ?string $from;
+
+    public function __construct(
+        #[\SensitiveParameter] string $apiKey,
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoTransport.php]:
```suggestion
        #[\SensitiveParameter]
string $apiKey,
```

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoTransport.php [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ * @author Samaël tomas <samael.tomas@gmail.com>
+ */
+final class PrimotextoTransport extends AbstractTransport
+{
+    protected const HOST = 'api.primotexto.com';
+
+    private string $apiKey;
+    private ?string $from;
+
+    public function __construct(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoTransport.php:31]: Please use CPP

File: src/Symfony/Component/Notifier/Bridge/Primotexto/README.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
+Primotexto Notifier
+=================
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/README.md]: Underline too short

File: src/Symfony/Component/Notifier/Bridge/Primotexto/README.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+DSN example
+-----------
+
+```
+PRIMOTEXTO_DSN=primotexto://APIKEY@default?from=FROM
+```
+
+where:
+ - `APIKEY` is your Primotexto API key
+ - `FROM` is your sender name
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/README.md]:
```suggestion
  - `APIKEY` is your Primotexto API key
  - `FROM` is your sender name
```

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $this->options['category'] = $category;
+
+        return $this;
+    }
+
+    /**
+     * @return $this
+     */
+    public function date(int $timestamp): static
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php]: What does date mean?

File: src/Symfony/Component/Notifier/Bridge/Primotexto/README.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+Resources
+---------
+
+* [Primotexto error codes](https://www.primotexto.com/api/plus/code_erreurs)
+
+ * [Contributing](https://symfony.com/doc/current/contributing/index.html)
+ * [Report issues](https://github.com/symfony/symfony/issues) and
+   [send Pull Requests](https://github.com/symfony/symfony/pulls)
+   in the [main Symfony repository](https://github.com/symfony/symfony)
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/README.md]:
```suggestion
 * [Primotexto error codes](https://www.primotexto.com/api/plus/code_erreurs)

 * [Contributing](https://symfony.com/doc/current/contributing/index.html)
 * [Report issues](https://github.com/symfony/symfony/issues) and
   [send Pull Requests](https://github.com/symfony/symfony/pulls)
   in the [main Symfony repository](https://github.com/symfony/symfony)
```

File: src/Symfony/Component/Notifier/Bridge/Primotexto/composer.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=8.2",
+        "symfony/http-client": "^6.4|^7.0",
+        "symfony/notifier": "^6.4|^7.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/composer.json]:
```suggestion
        "symfony/notifier": "^7.2"
```

File: src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php [hunk 13]
```diff
@@ -328,6 +328,10 @@ class UnsupportedSchemeException extends LogicException
             'class' => Bridge\Zulip\ZulipTransportFactory::class,
             'package' => 'symfony/zulip-notifier',
         ],
+        'primotexto' => [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php]: Please alpha sort

File: src/Symfony/Component/Notifier/Tests/Exception/UnsupportedSchemeExceptionTest.php [hunk 14]
```diff
@@ -103,6 +103,7 @@ public static function setUpBeforeClass(): void
             Bridge\Yunpian\YunpianTransportFactory::class => false,
             Bridge\Zendesk\ZendeskTransportFactory::class => false,
             Bridge\Zulip\ZulipTransportFactory::class => false,
+            Bridge\Primotexto\PrimotextoTransportFactory::class => false,
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Tests/Exception/UnsupportedSchemeExceptionTest.php]: Please alpha sort

File: src/Symfony/Component/Notifier/Tests/Exception/UnsupportedSchemeExceptionTest.php [hunk 15]
```diff
@@ -184,6 +185,7 @@ public static function messageWhereSchemeIsPartOfSchemeToPackageMapProvider(): \
         yield ['zendesk', 'symfony/zendesk-notifier'];
         yield ['zulip', 'symfony/zulip-notifier'];
         yield ['goip', 'symfony/go-ip-notifier'];
+        yield ['primotexto', 'symfony/primotexto-notifier'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Tests/Exception/UnsupportedSchemeExceptionTest.php]: Please alpha sort

File: src/Symfony/Component/Notifier/Transport.php [hunk 16]
```diff
@@ -105,6 +105,7 @@ final class Transport
         Bridge\Yunpian\YunpianTransportFactory::class,
         Bridge\Zendesk\ZendeskTransportFactory::class,
         Bridge\Zulip\ZulipTransportFactory::class,
+        Bridge\Primotexto\PrimotextoTransportFactory::class,
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Transport.php]: Please aloha sort

**Review by OskarStark — CHANGES_REQUESTED** [review]
Please update UnsupportedSchemeException and UnsupportedSchemeExceptionTest, thanks

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php (hunk 10 above)
Comment (Goazil) [reply src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php]: I added comment to explain and renaming the function to be more explicit

File: src/Symfony/Component/Notifier/Bridge/Primotexto/composer.json (hunk 12 above)
Comment (Goazil) [reply src/Symfony/Component/Notifier/Bridge/Primotexto/composer.json]: I specified `^7.1` instead

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoErrorCode.php [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    case INVALID_PHONE_NUMBER_SYNTAX = 11;
+    case NUMBER_BLACKLISTED_UNSUBSCRIBED = 12;
+    case NUMBER_BLACKLISTED_USER_BOUNCE = 13;
+    case NUMBER_BLACKLISTED_GLOBAL_BOUNCE = 14;
+    case NUMBER_ALREADY_EXISTS = 15;
+    case NO_MESSAGE_FOR_THIS_QUERY = 16;
+    case TOO_MANY_FIELDS_SELECTED = 17;
+    case FILE_TOO_LARGE = 18;
+    case PHONE_NUMBER_IS_FIXED_LINE = 19;
+
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoErrorCode.php]: No blank lines needed

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+namespace Symfony\Component\Notifier\Bridge\Primotexto;
+
+use Symfony\Component\Notifier\Message\MessageOptionsInterface;
+
+/**
+ * @author Samaël tomas <samael.tomas@gmail.com>
+ */
+final class PrimotextoOptions implements MessageOptionsInterface
+{
+    public function __construct(private array $options = [])
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php:21]:
```suggestion
    public function __construct(
private array $options = [])
```
Please linebreak

File: src/Symfony/Component/Notifier/Bridge/Primotexto/README.md [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+```php
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Bridge\Primotexto\options;
+
+$sms = new SmsMessage('+1411111111', 'My message');
+
+$options = (new PrimotextoOptions())
+    ->campaignName('Code de confirmation')
+    ->category('codeConfirmation')
+    ->date(1398177000000)
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/README.md]:
```suggestion
    -> campaignDate(1398177000000)
```
Wrong method name

File: src/Symfony/Component/Notifier/Bridge/Primotexto/README.md [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+Adding Options to a Message
+---------------------------
+
+With a Primotexto Message, you can use the `PrimotextoOptions` class to add
+[message options](https://www.primotexto.com/api/sms/notification.asp).
+
+```php
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Bridge\Primotexto\options;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/README.md]:
```suggestion
use Symfony\Component\Notifier\Bridge\Primotexto\options;
```
Wrong class

**Review by OskarStark — COMMENTED** [review]
Looks like your committer email is not associated with your GitHub account

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 21]
```diff
@@ -2185,9 +2185,9 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
                 ->setFactory([new Reference('messenger.transport_factory'), 'createTransport'])
                 ->setArguments([$transport['dsn'], $transport['options'] + ['transport_name' => $name], new Reference($serializerId)])
                 ->addTag('messenger.receiver', [
-                        'alias' => $name,
-                        'is_failure_transport' => \in_array($name, $failureTransports, true),
-                    ]
+                    'alias' => $name,
+                    'is_failure_transport' => \in_array($name, $failureTransports, true),
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Should be reverted

File: src/Symfony/Component/Notifier/Bridge/Primotexto/composer.json (hunk 12 above)
Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/Primotexto/composer.json]: Please use 7.2, as we are in feature freeze and this bridge will make it into 7.2

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoTransport.php [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SentMessage;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Samaël tomas <samael.tomas@gmail.com>
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoTransport.php]:
```suggestion
 * @author Samaël Tomas <samael.tomas@gmail.com>
```
Same for the others?

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    {
+    }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php]:
```suggestion
    public function __construct(
        private array $options = [],
    ){
    }
```

**Comment by Goazil** [comment]
Thanks for the review @OskarStark, I hope it's okay now :)

File: src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+use Symfony\Component\Notifier\Message\MessageOptionsInterface;
+
+/**
+ * @author Samaël Tomas <samael.tomas@gmail.com>
+ */
+final class PrimotextoOptions implements MessageOptionsInterface
+{
+    public function __construct(
+        private array $options = [],
+    ){
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Primotexto/PrimotextoOptions.php]:
```suggestion
    ) {
```

**Comment by OskarStark** [comment]
Can be merged into 7.2 if available

**Review by OskarStark — APPROVED** [review]
After my last small comment.

In the meantime you can please create a PR against `symfony/symfony-docs` and `symfony/recipes` repo, thanks

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Goazil.

**Pull request merged**